Fabric platform impl

emilyy-dev opened the pull request:
Closes #26

TODO:
* [x] Javadoc SOON:tm:

emilyy-dev committed:
Fabric platform impl

lucyydotp commented on gradlew.bat:
```diff
@@ -1,17 +1,24 @@
 @rem
-@rem Copyright 2015 the original author or authors.
+@rem Copyright © 2021 Lucy Poulton
```
I don't own copyright over this

emilyy-dev replied:
Oh shoot! Massive oversight! Reverted in https://github.com/lucyy-mc/Squirtgun/pull/29/commits/d0475edd98d18594eb7e3775f4ca0177f4d1852a :sweat_smile:

lucyydotp commented on squirtgun-api/src/main/java/me/lucyy/squirtgun/platform/Platform.java:
```diff
@@ -102,7 +102,7 @@
 	 * Gets a list of all online players.
 	 */
-	List<SquirtgunPlayer> getOnlinePlayers();
+	List<? extends SquirtgunPlayer> getOnlinePlayers();
```
what's this needed for?

emilyy-dev replied:
So in platform impls it can return things like `List<FabricPlayer>`, `List<BukkitPlayer>` etc.

lucyydotp replied:
why cant they just return a `List<SquirtgunPlayer>`? maybe im being thick but i dont get why this is necessary

emilyy-dev replied:
Actually this is useless, bounded wildcards are erased and compile down to just `T`. Will revert by the next commit probs

lucyydotp commented on squirtgun-platform-fabric/build.gradle.kts:
```diff
@@ -0,0 +1,102 @@
+
+            pom {
+                artifactId = artifactId
+                description.set("A multipurpose library designed for Minecraft: Java Edition plugins.")
```
could we maybe change this to something fabric-specific? i've been planning to do that with the other platforms too

emilyy-dev replied:
Sure

lucyydotp commented on squirtgun-platform-fabric/src/main/java/me/lucyy/squirtgun/fabric/FabricPlatform.java:
```diff
@@ -0,0 +1,214 @@
+	 * @param server server instance to work with
+	 */
+	public FabricPlatform(final @NotNull MinecraftServer server) {
+		this.server = validateState(MinecraftDedicatedServer.class::isInstance, server, MinecraftDedicatedServer.class::cast, "Squirtgun cannot be used on the client!");
```
why this restriction?

emilyy-dev replied:
For the most part it is not strictly necessary, but for something like getting the auth mode
https://github.com/lucyy-mc/Squirtgun/blob/7d9e580ab9d77a728aaa35a1e7f56360b5ce1239/squirtgun-platform-fabric/src/main/java/me/lucyy/squirtgun/fabric/FabricPlatform.java#L131
The "server properties" only exists on the [dedicated server](https://maven.fabricmc.net/docs/yarn-1.17+build.13/net/minecraft/server/dedicated/DedicatedServer.html#getProperties()), not the integrated one. The integrated server technically is always online since there is no `online-mode` option anywhere, let alone one connected to a proxy server. Can put a simple `instanceof` for that I guess? Not that it's an accurate measure.

lucyydotp replied:
ah alright thats fine, thanks for clarifying

lucyydotp commented on squirtgun-api/src/main/java/me/lucyy/squirtgun/platform/scheduler/Task.java:
```diff
@@ -128,4 +147,40 @@ public void execute(Platform platform) {
+
+	@Override
+	public int hashCode() {
+		int result = this.delay;
```
would Objects.hash not be easier?

emilyy-dev replied:
Maybe, it's auto-generated anyway and "[it just works](https://stackoverflow.com/a/1646913/8083725)", can be changed if requested

emilyy-dev replied:
Went ahead and changed it https://github.com/lucyy-mc/Squirtgun/pull/29/commits/87538d8c7838cbc7ddf8c27639c688dd2a6a6149

lucyydotp reviewed (changes requested):
just a few comments, overall looks great tho, tysm :D

emilyy-dev committed:
Revert modified license in gradlew scripts

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Some Task changes

Use Objects.hash for Task::hashCode
Null check in Task::equals to avoid npe

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Some rollbacks

Do not constain SG's usage solely to a dedicated server
Stupid type erasure

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Misc changes

Added DummyFabricPlayer to not return null on FabricPlatform::getPlayer
fabric-loom hates me, trying to fix our relationship xoxo

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Changes around FabricPlayer

Replace FabricPlayer with FabricPlayerImpl, make FabricPlayer an interface and add DummyFabricPlayer as an enum singleton

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Don't include Luck's Permission API

Just depend on it and suggest LuckPerms

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
cleaner

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Replace `var` usage with explicit types

Source jars look funny with `var`

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Add some javadoc; replace more `var` usages

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

lucyydotp commented on squirtgun-platform-fabric/src/main/java/me/lucyy/squirtgun/fabric/FabricNodeExecutor.java:
```diff
@@ -0,0 +1,91 @@
+	private final FabricPlatform platform;
+
+	public FabricNodeExecutor(final @NotNull CommandNode<PermissionHolder> commandNode,
+														final @NotNull FormatProvider formatProvider,
```
that's quite the indent you have there

emilyy-dev replied:
The tabs align in the IDE :p <s>this is why you use spaces instead</s>

lucyydotp reviewed (commented):
Thanks, this looks all good. How would I go about using this in a plugin implementation?

emilyy-dev commented:
> Thanks, this looks all good. How would I go about using this in a plugin implementation?

It isn't any harder than using it for Bukkit or Bungee really, I will be showing a quick example mod later.

emilyy-dev committed:
Fix (workaround?) fabric-loom's stupidity

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Fixes and refactoring

* Expose squirtgun-* dependencies so source jars & javadoc are linked too
* `FabricNodeExecutor`: use a `Supplier<FabricPlatform>` instead, see comment

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev committed:
Bugfix: Task$Builder::delay not checking correct variable

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev commented:
@lucyy-mc I've made a simple sample mod for it over at [emilyy-dev/squirtgun-fabric-sample-mod](https://github.com/emilyy-dev/squirtgun-fabric-sample-mod) :)

lucyy-mc commented:
thank you, i'll have a play with this

lucyydotp reviewed (commented):
lgtm

lucyy-mc commented:
```
> Configure project :squirtgun-platform-fabric
Fabric Loom: 0.8.21
java.nio.file.NoSuchFileException: /home/lucy/src/sg-fabric/squirtgun-commands/build/libs/squirtgun-commands-2.0.0-pre6-SNAPSHOT.jar
```

I keep on getting this error trying to build clean, can you reproduce this?

emilyy-dev commented:
> ```
> > Configure project :squirtgun-platform-fabric
> Fabric Loom: 0.8.21
> java.nio.file.NoSuchFileException: /home/lucy/src/sg-fabric/squirtgun-commands/build/libs/squirtgun-commands-2.0.0-pre6-SNAPSHOT.jar
> ```
> 
> I keep on getting this error trying to build clean, can you reproduce this?

Actually now that it's brought up yes, I do remember working around that in the past, will look into it!

emilyy-dev committed:
Make this work once and for all!

Signed-off-by: emilyy-dev <35617540+emilyy-dev@users.noreply.github.com>

emilyy-dev commented:
I believe this is ready for a final review and hopefully merged!
Also updated the demo mod according to a decision made, see https://github.com/emilyy-dev/squirtgun-fabric-sample-mod/commit/0818ed942dc3d485222246dc8c19156e53906c62
